JavaScript/JavaScriptSDK/ajax/ajax.ts

```diff
@@ -291,211 +58,284 @@ module Microsoft.ApplicationInsights {
+                                    (<any>this).ajaxData.callbackFinishedTime = dateTime.Now();
                                     }
-                                });
                             }
 
-                            onReadyStateChangePostfix.call(this);
+                            ajaxMonitorInstance.onReadyStateChangePostfix(xhr);
                         }
                     }
-                };
+                } catch (e) {
+                    _InternalLogging.throwInternalNonUserActionable(
+                        LoggingSeverity.CRITICAL,
+                        "Failed to monitor XMLHttpRequest.onreadystatechange, monitoring data for this ajax call may be incorrect: "
+                        + Microsoft.ApplicationInsights.Util.dump(e));
+                }
+                        }
+            (<any>wrapper).instrumentedByAppInsights = true;
+
+            return wrapper;
+                    }
 
-                function onReadStateChangePrefix() {
-                    switch (this.readyState) {
+        private onReadStateChangePrefix(xhr: XMLHttpRequest) {
+            switch (xhr.readyState) {
                         case 3:
-                            commands.TryCatchTraceWrapper.call(this, "readyState(3)", function () {
-                                this.ajaxData.responseStartedTime = dateTime.Now();
-                            });
+                    (<any>xhr).ajaxData.responseStartedTime = dateTime.Now();
                             break;
                         case 4:
-                            commands.TryCatchTraceWrapper.call(this, "readyState(4)", collectResponseData);
+                    this.collectResponseData(xhr);
                             break;
                     }
                 }
 
-                function onReadyStateChangePostfix() {
-                    if (this.readyState === 4) {
-                        onAjaxComplete.call(this);
+        private onReadyStateChangePostfix(xhr: XMLHttpRequest) {
+            if (xhr.readyState === 4) {
+                this.onAjaxComplete(xhr);
                     }
                 }
 
-                function onAjaxComplete() {
-                    commands.TryCatchTraceWrapper.call(this, "publishData", function () {
-                        this.ajaxData.CalculateMetrics();
+        private onAjaxComplete(xhr: XMLHttpRequest) {
+            try {
+                (<any>xhr).ajaxData.CalculateMetrics();
 
                         var successStatuses = [200, 201, 202, 203, 204, 301, 302, 303, 304];
 
-                        appInsights.trackAjax(
-                            this.ajaxData.getAbsoluteUrl(),
-                            this.ajaxData.async,
-                            this.ajaxData.ajaxTotalDuration,
-                            successStatuses.indexOf(+this.ajaxData.status) != -1
+                this.appInsights.trackAjax(
+                    (<any>xhr).ajaxData.getAbsoluteUrl(),
```

Comment:
You'll start getting `OtherValueReservedforAIUser` instead of the name in charts and sausages